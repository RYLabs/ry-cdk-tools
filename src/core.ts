export * from "./base_stacks/base_stack";
export { default as BaseStack } from "./base_stacks/base_stack";
export { default as Conventions } from "./utils/conventions";
