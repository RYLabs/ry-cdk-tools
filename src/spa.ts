export { default as SpaPipelineStack } from "./stacks/spa_pipeline_stack";
export * from "./stacks/spa_pipeline_stack";
