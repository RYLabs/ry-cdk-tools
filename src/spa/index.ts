export { default as S3SpaPipelineStack } from "../stacks/s3_spa_pipeline_stack";
export * from "../stacks/s3_spa_pipeline_stack";
export { default as S3Spa } from "../spa/s3_spa";
export * from "../spa/s3_spa";
export { default as CloudFrontConfig } from "./cloudfront_config";
export * from "./cloudfront_config";
