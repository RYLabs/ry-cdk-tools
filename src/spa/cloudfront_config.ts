import { Construct, CfnOutput } from '@aws-cdk/core';
import { IBucket } from '@aws-cdk/aws-s3';
import { IHostedZone, ARecord, RecordTarget } from '@aws-cdk/aws-route53';
import { CloudFrontTarget } from '@aws-cdk/aws-route53-targets/lib';
import {
  CloudFrontWebDistribution,
  SSLMethod,
  SecurityPolicyProtocol
} from '@aws-cdk/aws-cloudfront';

export interface CloudFrontConfigProps {
  subDomain?: string;
  /**
   * certificateArn: "arn:aws:acm:region:123456789012:certificate/12345678-1234-1234-1234-123456789012"
   * This value can be entered manually from your own certficate stored with ACM
   * or you can create Route53AcmSslStack.  
   */
  certificateArn: string;
  /**
   *  sourceBucket: IBucket
   *  The S3 bucket where your live project is located.
   */
  sourceBucket: IBucket;
  /**
   *  hostedZone: IHostedZone
   *  Route53 HostedZone created via Route53AcmSslStack
   */
  hostedZone: IHostedZone;
}

export default class CloudFrontConfig extends Construct {
  constructor(scope: Construct, id: string, props: CloudFrontConfigProps) {
    super(scope, id);

    const {
      subDomain = "www",
      certificateArn,
      sourceBucket,
      hostedZone,
    } = props;

    const fullDomain = subDomain + "." + hostedZone.zoneName;
    
    const distribution = new CloudFrontWebDistribution(this, `projectDistribution`, {
      aliasConfiguration: {
        acmCertRef: certificateArn,
        names: [fullDomain],
        sslMethod: SSLMethod.SNI,
        securityPolicy: SecurityPolicyProtocol.TLS_V1_2_2018
      },
      originConfigs: [
        {
          s3OriginSource: {
            s3BucketSource: sourceBucket,
          },
          behaviors: [ {isDefaultBehavior: true} ]
        },
      ],
    });

    new ARecord(this, 'SiteAliasRecord', {
      recordName: fullDomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
      zone: hostedZone,
    });

    new CfnOutput(this, 'Project HTTPS URL', {
      value: 'https://' + fullDomain,
      description: 'Project URL using CloudFront Route53 for SSL'
    });
  }
}
