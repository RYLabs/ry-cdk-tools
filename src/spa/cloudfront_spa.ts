import { Construct, CfnOutput } from "@aws-cdk/core";
import { IBucket } from "@aws-cdk/aws-s3";
import {
  IHostedZone,
  ARecord,
  RecordTarget,
  HostedZone,
} from "@aws-cdk/aws-route53";
import { CloudFrontTarget } from "@aws-cdk/aws-route53-targets/lib";
import { DnsValidatedCertificate } from "@aws-cdk/aws-certificatemanager";
import S3Spa from "./s3_spa";
import { CloudFrontWebDistribution } from "@aws-cdk/aws-cloudfront";

export interface CloudfrontSpaProps {
  subDomain?: string;
  /**
   * certificateArn: "arn:aws:acm:region:123456789012:certificate/12345678-1234-1234-1234-123456789012"
   * This value can be entered manually from your own certficate stored with ACM
   * or you can create Route53AcmSslStack.
   */
  certificateArn?: string;
  /**
   *  sourceBucket: IBucket
   *  The S3 bucket where your live project is located.
   *  If no bucket is provided then a new bucket is created.
   */
  sourceBucket?: IBucket;
  domainName: string;
}

function createCert(scope: any, domainName: string, zone: IHostedZone) {
  const cert = new DnsValidatedCertificate(scope, "projectCertificate", {
    domainName: domainName,
    hostedZone: zone,
  });
  return cert.certificateArn;
}

function createBucket(scope: any) {
  const bucket = new S3Spa(scope, "projectBucket", {
    bucketName: scope.conventions.eqn("dash").toLowerCase(),
  });
  return bucket;
}

export default class CloudfrontSpa extends Construct {
  constructor(scope: Construct, id: string, props: CloudfrontSpaProps) {
    super(scope, id);

    const {
      subDomain = "www",
      certificateArn,
      sourceBucket,
      domainName,
    } = props;

    const hostedZone = HostedZone.fromLookup(this, "projectHostedZone", {
      domainName: domainName,
    });

    const fullDomain = subDomain + "." + hostedZone.zoneName;

    const distribution = new CloudFrontWebDistribution(
      this,
      `projectDistribution`,
      {
        aliasConfiguration: {
          acmCertRef:
            certificateArn || createCert(this, domainName, hostedZone),
          names: [fullDomain, domainName],
        },
        originConfigs: [
          {
            s3OriginSource: {
              s3BucketSource: sourceBucket || createBucket(this).bucket,
            },
            behaviors: [{ isDefaultBehavior: true }],
          },
        ],
      }
    );

    new ARecord(this, "SiteAliasRecord", {
      recordName: fullDomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
      zone: hostedZone,
    });

    new CfnOutput(this, "Project HTTPS URL", {
      value: "https://" + fullDomain,
      description: "Project URL using CloudFront Route53 for SSL",
    });
  }
}
