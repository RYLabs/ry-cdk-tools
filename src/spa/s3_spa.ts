import { Construct, RemovalPolicy } from "@aws-cdk/core";
import { Bucket } from "@aws-cdk/aws-s3";

export interface S3SpaProps {
  bucketName: string;
}

// Single page application using S3
export default class S3Spa extends Construct {
  bucket: Bucket;

  get websiteUrl(): string {
    return this.bucket.bucketWebsiteUrl;
  }

  // Construct needed to be passed for the creation of resources.
  constructor(scope: Construct, id: string, props: S3SpaProps) {
    super(scope, id);

    const { bucketName } = props;

    this.bucket = new Bucket(this, "siteBucket", {
      bucketName,
      websiteIndexDocument: "index.html",

      // re-render the index doc to support pushState
      websiteErrorDocument: "index.html",

      publicReadAccess: true,
      removalPolicy: RemovalPolicy.DESTROY,
    });
  }
}
