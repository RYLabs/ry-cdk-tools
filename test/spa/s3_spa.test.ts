import { SynthUtils } from "@aws-cdk/assert";
import { Stack } from "@aws-cdk/core";

import { S3Spa } from "../../lib/spa";

test("creates a single page application", () => {
  const stack = new Stack();
  new S3Spa(stack, "spa", {
    bucketName: "test-site",
  });
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
