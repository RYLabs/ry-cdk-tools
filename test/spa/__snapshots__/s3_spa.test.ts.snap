// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a single page application 1`] = `
Object {
  "Resources": Object {
    "spasiteBucket0C2DE9A5": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketName": "test-site",
        "WebsiteConfiguration": Object {
          "ErrorDocument": "index.html",
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "spasiteBucketPolicy0093B613": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "spasiteBucket0C2DE9A5",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": "*",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "spasiteBucket0C2DE9A5",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;
